File: services/modules/domain/src/main/scala/solutions/s4y/vocabla/domain/TagSmall.scala
package solutions.s4y.vocabla.domain

import solutions.s4y.vocabla.domain.identity.{
  Identified,
  Identifier,
  IdentifierSchema
}
import zio.json.{DeriveJsonCodec, JsonCodec}
import zio.prelude.Equal
import zio.schema.{Schema, derived}

final case class TagSmall(
    label: String
):
  override def toString: String = s"Tag(small): $label"

object TagSmall:
  given Equal[Tag] =
    Equal.make((a, b) => a.label == b.label)

  given Equal[Identified[Tag]] =
    Equal.make((a, b) => a.e.label == b.e.label)

  given (using is: IdentifierSchema): Schema[TagSmall] = Schema.derived
  given (using IdentifierSchema): JsonCodec[TagSmall] =
    DeriveJsonCodec.gen[TagSmall]

File: front_react/src/domain/TagSmall.ts
import {type Identifier} from './identity/Identifier.ts';
import {Schema} from 'effect';

export type TagSmall = {
  label: string,
}

export const Tag = (label: string): TagSmall => ({label})

export type TagSmallId = Identifier<TagSmall>

export const schemaTagSmall = Schema.Struct({
  label: Schema.String,
})
